File: server/Mappers/ProductMapper.cs
using server.DTOs;
using server.Models;
using System.Linq;
using System.Collections.Generic;

namespace server.Mappers
{
    public static class ProductMapper
    {
        public static ProductListDto ToProductListDto(this Product p)
        {
            return new ProductListDto
            {
                Id = p.Id,
                Title = p.Title,
                Description = p.Description,
                ImageUrl = p.ImageUrl,
                CreationDate = p.CreationDate,
                CreatorUserId = p.CreatorUserId
            };
        }

        public static ProductDto ToProductDto(this Product p)
        {
            return new ProductDto
            {
                Id = p.Id,
                Title = p.Title,
                Description = p.Description,
                ImageUrl = p.ImageUrl,
                CreationDate = p.CreationDate,
                CreatorUserId = p.CreatorUserId,
                ProductCategoryId = p.ProductCategoryId,
                Categories = p.ProductCategories?.Select(c => c.Id).ToList() ?? new List<int>(),
                Comments = p.Comments?.Where(c => !c.IsDeleted).Select(c => new CommentDto
                {
                    Id = c.Id,
                    Description = c.Description,
                    CreationDate = c.CreationDate
                }).ToList() ?? new List<CommentDto>()
            };
        }

        public static Product FromProductsListDto(this ProductDto dto)
        {
            return new Product
            {
                Id = dto.Id,
                Title = dto.Title,
                Description = dto.Description,
                ImageUrl = dto.ImageUrl,
                CreationDate = dto.CreationDate,
                CreatorUserId = dto.CreatorUserId,
                ProductCategoryId = dto.ProductCategoryId
            };
        }
    }
}

File: client/src/mappers/ProductMapper.ts
import { ProductDetailDto } from "@/DTOs/ProductDto";
import { ProductListDto } from "@/DTOs/ProductDto";
import { mapToCommentListDto } from "./CommentMapper";

export const mapToProductDto=(rawProduct : any): ProductDetailDto=>{

    return{
        id: rawProduct.id ||-1,
        title: rawProduct.title || "Brak nazwy",
        description:rawProduct.description || '',
        imageUrl: rawProduct.imageUrl || null,
        categoryId: rawProduct.productCategoryId || -1,
        creationDate: rawProduct.creationDate ? new Date(rawProduct.creationDate) : null
        ,creatorUserId:rawProduct.creatorUserId || -1,
          
        comments: mapToCommentListDto(rawProduct.comments),
        productCategoryList: Array.isArray(rawProduct.categories) ? rawProduct.categories : [],
        productCategoryIdList: Array.isArray(rawProduct.categoriesIds) ? rawProduct.categoriesIds : [],
          }; 
           
}

export const mapToProductList=(rawProduct : any): ProductListDto=>{

    return{
        id: rawProduct.id ||-1,
        title: rawProduct.title || "Brak nazwy",
        description:rawProduct.description || '',
        imageUrl: rawProduct.imageUrl || null,
        categoryId: rawProduct.productCategoryId || -1,
        creationDate: rawProduct.creationDate ? new Date(rawProduct.creationDate) : null
        ,creatorUserId:rawProduct.creatorUserId || -1,
          };  
}

export const mapToProductListDto=(rawInput:any):ProductListDto[]=>{
    if (!Array.isArray(rawInput)){
        return [];
    };

    return rawInput.map((product)=>mapToProductList(product))

  }
